File: netrail/backends/brave.py
from __future__ import annotations

import logging
import os

import httpx

from netrail.backends.types import OPERATORS, SearchMode, SearchResult

logger = logging.getLogger(__name__)

PROVENANCE = "Brave Search API (your key, your quota)"


class BraveBackend:
    name = "brave"
    provenance = PROVENANCE
    supports_operators = OPERATORS

    def __init__(self, api_key: str, timeout: float = 12.0) -> None:
        self.api_key = api_key
        self.timeout = timeout

    @classmethod
    def from_env(cls) -> "BraveBackend | None":
        return cls.from_env_var(None)

    @classmethod
    def from_env_var(cls, env_name: str | None) -> "BraveBackend | None":
        primary = env_name or "BRAVE_SEARCH_API_KEY"
        candidates = [primary]
        if primary != "NETRAIL_BRAVE_API_KEY":
            candidates.append("NETRAIL_BRAVE_API_KEY")
        if primary != "BRAVE_SEARCH_API_KEY":
            candidates.append("BRAVE_SEARCH_API_KEY")
        key = next((os.environ.get(name) for name in candidates if os.environ.get(name)), None)
        if not key or not key.strip():
            return None
        return cls(key.strip())

    def is_available(self) -> bool:
        return bool(self.api_key)

    def search(self, query: str, mode: SearchMode, max_results: int) -> list[SearchResult]:
        if mode == "images":
            return self._search_images(query, max_results)
        return self._search_web(query, max_results)

    def _search_web(self, query: str, max_results: int) -> list[SearchResult]:
        headers = {
            "Accept": "application/json",
            "X-Subscription-Token": self.api_key,
        }
        params = {"q": query, "count": min(max_results, 20)}
        with httpx.Client(timeout=self.timeout) as client:
            response = client.get(
                "https://api.search.brave.com/res/v1/web/search",
                headers=headers,
                params=params,
            )
            response.raise_for_status()
            payload = response.json()

        results: list[SearchResult] = []
        for item in payload.get("web", {}).get("results", [])[:max_results]:
            url = item.get("url", "")
            if not url:
                continue
            results.append(
                SearchResult(
                    title=item.get("title") or url,
                    url=url,
                    snippet=item.get("description", ""),
                    backend=self.name,
                    provenance=self.provenance,
                )
            )
        return results

    def _search_images(self, query: str, max_results: int) -> list[SearchResult]:
        headers = {
            "Accept": "application/json",
            "X-Subscription-Token": self.api_key,
        }
        params = {"q": query, "count": min(max_results, 20)}
        with httpx.Client(timeout=self.timeout) as client:
            response = client.get(
                "https://api.search.brave.com/res/v1/images/search",
                headers=headers,
                params=params,
            )
            response.raise_for_status()
            payload = response.json()

        results: list[SearchResult] = []
        for item in payload.get("results", [])[:max_results]:
            url = item.get("url", "")
            if not url:
                continue
            thumb = item.get("thumbnail") or {}
            results.append(
                SearchResult(
                    title=item.get("title") or "Image result",
                    url=url,
                    snippet=item.get("source", ""),
                    image=thumb.get("src"),
                    source=item.get("source", ""),
                    backend=self.name,
                    provenance=self.provenance,
                )
            )
        return results

File: src-tauri/src/backends/brave.rs
use super::types::{SearchMode, SearchResult};
use crate::error::{NetRailError, NetRailResult};
use reqwest::Client;
use std::env;

pub const PROVENANCE: &str = "Brave Search API (your key, your quota)";

pub struct BraveBackend {
    api_key: String,
    client: Client,
}

impl BraveBackend {
    pub fn from_env(client: Client) -> Option<Self> {
        Self::from_env_var(client, None)
    }

    pub fn from_env_var(client: Client, env_name: Option<&str>) -> Option<Self> {
        let primary = env_name.unwrap_or("BRAVE_SEARCH_API_KEY");
        let key = env::var(primary)
            .or_else(|_| {
                if primary != "NETRAIL_BRAVE_API_KEY" {
                    env::var("NETRAIL_BRAVE_API_KEY")
                } else {
                    Err(env::VarError::NotPresent)
                }
            })
            .or_else(|_| {
                if primary != "BRAVE_SEARCH_API_KEY" {
                    env::var("BRAVE_SEARCH_API_KEY")
                } else {
                    Err(env::VarError::NotPresent)
                }
            })
            .ok()?;
        if key.trim().is_empty() {
            return None;
        }
        Some(Self {
            api_key: key,
            client,
        })
    }

    pub fn name(&self) -> &'static str {
        "brave"
    }

    pub fn provenance(&self) -> &'static str {
        PROVENANCE
    }

    pub fn is_available(&self) -> bool {
        !self.api_key.is_empty()
    }

    pub async fn search(
        &self,
        query: &str,
        mode: SearchMode,
        max_results: usize,
    ) -> NetRailResult<Vec<SearchResult>> {
        match mode {
            SearchMode::Images => self.search_images(query, max_results).await,
            SearchMode::Web => self.search_web(query, max_results).await,
        }
    }

    async fn search_web(&self, query: &str, max_results: usize) -> NetRailResult<Vec<SearchResult>> {
        let response = self
            .client
            .get("https://api.search.brave.com/res/v1/web/search")
            .header("Accept", "application/json")
            .header("X-Subscription-Token", &self.api_key)
            .query(&[("q", query), ("count", &max_results.min(20).to_string())])
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(NetRailError::BackendHttp {
                code: "BRAVE_HTTP_ERROR",
                backend: "brave".into(),
                status: response.status().as_u16(),
            });
        }

        let payload: serde_json::Value = response.json().await?;
        let mut results = Vec::new();

        if let Some(items) = payload
            .get("web")
            .and_then(|w| w.get("results"))
            .and_then(|v| v.as_array())
        {
            for item in items.iter().take(max_results) {
                let url = item.get("url").and_then(|v| v.as_str()).unwrap_or("");
                if url.is_empty() {
                    continue;
                }
                results.push(SearchResult {
                    title: item
                        .get("title")
                        .and_then(|v| v.as_str())
                        .unwrap_or(url)
                        .to_string(),
                    url: url.to_string(),
                    snippet: item
                        .get("description")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string(),
                    image: None,
                    source: "brave".into(),
                    backend: self.name().into(),
                    provenance: self.provenance().into(),
                });
            }
        }
        Ok(results)
    }

    async fn search_images(
        &self,
        query: &str,
        max_results: usize,
    ) -> NetRailResult<Vec<SearchResult>> {
        let response = self
            .client
            .get("https://api.search.brave.com/res/v1/images/search")
            .header("Accept", "application/json")
            .header("X-Subscription-Token", &self.api_key)
            .query(&[("q", query), ("count", &max_results.min(20).to_string())])
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(NetRailError::BackendHttp {
                code: "BRAVE_HTTP_ERROR",
                backend: "brave".into(),
                status: response.status().as_u16(),
            });
        }

        let payload: serde_json::Value = response.json().await?;
        let mut results = Vec::new();

        if let Some(items) = payload.get("results").and_then(|v| v.as_array()) {
            for item in items.iter().take(max_results) {
                let url = item.get("url").and_then(|v| v.as_str()).unwrap_or("");
                if url.is_empty() {
                    continue;
                }
                results.push(SearchResult {
                    title: item
                        .get("title")
                        .and_then(|v| v.as_str())
                        .unwrap_or("Image result")
                        .to_string(),
                    url: url.to_string(),
                    snippet: item
                        .get("source")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string(),
                    image: item
                        .get("thumbnail")
                        .and_then(|v| v.get("src"))
                        .and_then(|v| v.as_str())
                        .map(str::to_string),
                    source: item
                        .get("source")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string(),
                    backend: self.name().into(),
                    provenance: self.provenance().into(),
                });
            }
        }
        Ok(results)
    }
}